linkage_dag.py

```diff
@@ -353,25 +355,41 @@
         f"/snap/bin/gsutil -m cp -r gs://{bucket}/{gcs_folder}/simhash_results .",
         f"/snap/bin/gsutil -m cp -r gs://{bucket}/{tmp_dir}/prev_id_mapping .",
         "mkdir new_simhash_indexes",
-        (
-            "python3 run_simhash.py simhash_input simhash_results --simhash_indexes simhash_indexes "
-            "--new_simhash_indexes new_simhash_indexes"
-        ),
-        "cp -r article_pairs usable_ids",
-        "cp simhash_results/* article_pairs/",
-        (
-            "python3 create_merge_ids.py --match_dir usable_ids --prev_id_mapping_dir prev_id_mapping "
-            "--merge_file id_mapping.jsonl --current_ids_dir article_pairs"
-        ),
-        f"/snap/bin/gsutil -m cp id_mapping.jsonl gs://{bucket}/{gcs_folder}/tmp/",
-        f"/snap/bin/gsutil -m cp simhash_results/* gs://{bucket}/{gcs_folder}/simhash_results/",
-        f"/snap/bin/gsutil -m cp new_simhash_indexes/* gs://{bucket}/{gcs_folder}/simhash_indexes/",
     ]
-    vm_script = " && ".join(vm_script_sequence)
+    prep_environment_vm_script = " && ".join(prep_environment_script_sequence)
+
+    prep_environment = BashOperator(
+        task_id="prep_environment",
+        bash_command=f'gcloud compute ssh jm3312@{gce_resource_id} --zone {gce_zone} --command "{prep_environment_vm_script}"',
+    )
+
+    update_simhash_index = BashOperator(
+        task_id="update_simhash_index",
+        bash_command=f'gcloud compute ssh jm3312@{gce_resource_id} --zone {gce_zone} --command "bash run_simhash_scripts.sh &> log &"',
+    )
+
+    wait_for_simhash_index = GCSObjectExistenceSensor(
+        task_id="wait_for_simhash_index",
+        bucket=DATA_BUCKET,
+        object=f"{tmp_dir}/done_files/simhash_is_done",
+        deferrable=True,
+    )
 
     create_cset_ids = BashOperator(
         task_id="create_cset_ids",
-        bash_command=f'gcloud compute ssh jm3312@{gce_resource_id} --zone {gce_zone} --command "{vm_script}"',
+        bash_command=f'gcloud compute ssh jm3312@{gce_resource_id} --zone {gce_zone} --command "bash run_ids_scripts.sh &> log &"',
+    )
+
+    wait_for_cset_ids = GCSObjectExistenceSensor(
+        task_id="wait_for_cset_ids",
+        bucket=DATA_BUCKET,
+        object=f"{tmp_dir}/done_files/ids_are_done",
+        deferrable=True,
+    )
+
+    push_to_gcs = BashOperator(
+        task_id="push_to_gcs",
+        bash_command=f'gcloud compute ssh jm3312@{gce_resource_id} --zone {gce_zone} --command "run_ids_script.sh &"',
```

Comment:
Where is the `run_ids_script.sh` file? I couldn't find this one.

Reply:
It's in `utils`!

Reply:
Oh wait, I think you're right and there's a missing 's'. Thanks!